core/src/main/java/kafka/server/share/ShareSessionContext.java

```diff
@@ -0,0 +1,243 @@
+
+    public Map<TopicIdPartition, ShareFetchRequest.SharePartitionData> shareFetchData() {
+        return shareFetchData;
+    }
+
+    public boolean isSubsequent() {
+        return isSubsequent;
+    }
+
+    public ShareSession session() {
+        return session;
+    }
+
+    @Override
+    ShareFetchResponse throttleResponse(int throttleTimeMs) {
+        if (!isSubsequent) {
+            return new ShareFetchResponse(ShareFetchResponse.toMessage(Errors.NONE, throttleTimeMs,
+                    Collections.emptyIterator(), Collections.emptyList()));
+        } else {
+            int expectedEpoch = ShareFetchMetadata.nextEpoch(reqMetadata.epoch());
+            int sessionEpoch;
+            synchronized (session) {
+                sessionEpoch = session.epoch;
+            }
+            if (sessionEpoch != expectedEpoch) {
+                log.debug("Subsequent share session {} expected epoch {}, but got {}. " +
+                        "Possible duplicate request.", session.key(), expectedEpoch, sessionEpoch);
+                return new ShareFetchResponse(ShareFetchResponse.toMessage(Errors.INVALID_SHARE_SESSION_EPOCH,
+                        throttleTimeMs, Collections.emptyIterator(), Collections.emptyList()));
+            } else
+                return new ShareFetchResponse(ShareFetchResponse.toMessage(Errors.NONE, throttleTimeMs,
+                        Collections.emptyIterator(), Collections.emptyList()));
+        }
+    }
+
+    // Iterator that goes over the given partition map and selects partitions that need to be included in the response.
+    // If updateShareContextAndRemoveUnselected is set to true, the share context will be updated for the selected
+    // partitions and also remove unselected ones as they are encountered.
+    private class PartitionIterator implements
+            Iterator<Entry<TopicIdPartition, PartitionData>> {
+        private final Iterator<Map.Entry<TopicIdPartition, ShareFetchResponseData.PartitionData>> iterator;
+        private final boolean updateShareContextAndRemoveUnselected;
+        private Map.Entry<TopicIdPartition, ShareFetchResponseData.PartitionData> nextElement;
+
+
+        public PartitionIterator(Iterator<Map.Entry<TopicIdPartition, ShareFetchResponseData.PartitionData>> iterator, boolean updateShareContextAndRemoveUnselected) {
+            this.iterator = iterator;
+            this.updateShareContextAndRemoveUnselected = updateShareContextAndRemoveUnselected;
+        }
+
+        @Override
+        public boolean hasNext() {
+            while ((nextElement == null) && iterator.hasNext()) {
+                Map.Entry<TopicIdPartition, ShareFetchResponseData.PartitionData> element = iterator.next();
+                TopicIdPartition topicPart = element.getKey();
+                ShareFetchResponseData.PartitionData respData = element.getValue();
+                CachedSharePartition cachedPart = session.partitionMap().find(new CachedSharePartition(topicPart));
```

Comment:
Is it guaranteed that `cachedPart` can never be null?

Reply:
yes, since `cachedPart` will always be present as a part of share session's cached partition map.